Telcodocs 419: Updating Sysctl and creating section on setting interface sysctl values (#45888)
openshift/openshift-docs: modules/nodes-containers-start-pod-safe-sysctls.adoc

> // Module included in the following assemblies:

Reviewer: This pertains to the previous image...although this is a 'stand-in' image until the final one is ready, will 'plugin' be listed as 'plug-in'?  [ISG Word usage section: https://www.ibm.com/docs/en/ibm-style?topic=word-usage#p].

//
// * nodes/containers/nodes-containers-sysctls.adoc

:_content-type: PROCEDURE
[id="nodes-starting-pod-safe-sysctls_{context}"]
= Starting a pod with safe sysctls

> You can set sysctls on pods using the pod's `securityContext`. The `securityContext` applies to all containers in the same pod.

Reviewer 2: You can set sysctls on pods using the pod's 
=>
Sysctl's can be set ....

Author: funny my linter on the .adoc file complains as this is passive voice. We try to maintain active voice.

Safe sysctls are allowed by default.

This example uses the pod `securityContext` to set the following safe sysctls:

* `kernel.shm_rmid_forced`
* `net.ipv4.ip_local_port_range`
* `net.ipv4.tcp_syncookies`
* `net.ipv4.ping_group_range`

[WARNING]
====
To avoid destabilizing your operating system, modify sysctl parameters only after you understand their effects.
====

Use this procedure to start a pod with the configured sysctl settings.
[NOTE]
====

> In most cases you will be modifying an existing pod definition and adding the `securityContext` spec.

Reviewer 3: ```suggestion
In most cases, you modify an existing pod definition and adding the `securityContext` spec.
```

====

> .Procedure

Reviewer 3: Same not about user-replaceable values applies to this topic.

. Create a YAML file `sysctl_pod.yaml` that defines an example pod and add the `securityContext` spec, as shown in the following example:
+
[source,yaml]
----
apiVersion: v1
kind: Pod
metadata:
  name: sysctl-example
  namespace: default
spec:
  containers:
  - name: podexample
    image: centos
    command: ["bin/bash", "-c", "sleep INF"]
  securityContext:
    sysctls:
    - name: kernel.shm_rmid_forced
      value: "1"
    - name: net.ipv4.ip_local_port_range
      value: "32770       60666"
    - name: net.ipv4.tcp_syncookies
      value: "0"
    - name: net.ipv4.ping_group_range
      value: "0           200000000"
----

> +

Reviewer 3: Should this have callouts?

[NOTE]
====

> An example pod is created here simply to illustrate how the changes are applied.

Reviewer: ```suggestion
An example pod is created here to illustrate how the changes are applied.
```

====

> . Create the pod:

Reviewer 3: ```suggestion
. Create the pod by running the following command:
```

+
[source,terminal]
----

> $ oc apply -f sysctl_pod.yaml

Reviewer 4: When I apply sysctl_pod.yaml manifest the following warning appears:
```
W0616 18:00:31.447102 1349418 warnings.go:70] would violate PodSecurity "restricted:latest": allowPrivilegeEscalation != false (container "podexample" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "podexample" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "podexample" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "podexample" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
pod/sysctl-example created
```
It happens because of the latest changes in ocp4.11 security.
In order to suppers this warning the following yaml should be applied 

```
apiVersion: v1
kind: Pod
metadata:
  name: sysctl-example
  namespace: default
spec:
  containers:
  - name: podexample
    image: centos
    command: ["bin/bash", "-c", "sleep INF"]
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop: ["ALL"]
  securityContext:
    runAsNonRoot: true
    seccompProfile:
      type: RuntimeDefault
    sysctls:
    - name: kernel.shm_rmid_forced
      value: "1"
    - name: net.ipv4.ip_local_port_range
      value: "32770       60666"
    - name: net.ipv4.tcp_syncookies
      value: "0"
    - name: net.ipv4.ping_group_range
      value: "0           200000000"
```

Author: Thanks @kononovn I did notice that, but I thought it was something specific to the environment I tested on. Thanks for letting me know it is something new in 4.11

----
+

> . Verify that the pod is created:

Reviewer 3: ```suggestion
. Verify that the pod is created by running the following command:
```

+
[source,terminal]
----
$ oc get pod
----
+
.Example output
[source,terminal]
----
NAME              READY   STATUS            RESTARTS   AGE
sysctl-example    1/1     Running           0          14s
----

> . Log in to the pod:

Reviewer 3: ```suggestion
. Log in to the pod by running the following command:
```

+
[source,terminal]
----
$ oc rsh sysctl-example
----

> . Verify the values of the configured sysctl flags. For example find the value `kernel.shm_rmid_forced` as follows:

Reviewer 3: ```suggestion
. Verify the values of the configured sysctl flags. For example, find the value `kernel.shm_rmid_forced` by running the following command:
```

+
[source,terminal]
----

> sh-4.4# sysctl -a | grep kernel.shm_rmid_forced

Reviewer 4: When I exec this command in unprivileged pod I receive the following output
```
> oc rsh sysctl-example
sh-4.4$ sysctl -a | grep kernel.shm_rmid_forced
sysctl: permission denied on key 'fs.protected_fifos'
sysctl: permission denied on key 'fs.protected_hardlinks'
sysctl: permission denied on key 'fs.protected_regular'
sysctl: permission denied on key 'fs.protected_symlinks'
sysctl: permission denied on key 'kernel.cad_pid'
sysctl: permission denied on key 'kernel.usermodehelper.bset'
sysctl: permission denied on key 'kernel.usermodehelper.inheritable'
kernel.shm_rmid_forced = 1
sysctl: permission denied on key 'net.ipv4.tcp_fastopen_key'
sysctl: permission denied on key 'net.ipv6.conf.all.stable_secret'
sysctl: permission denied on key 'net.ipv6.conf.default.stable_secret'
sysctl: permission denied on key 'net.ipv6.conf.eth0.stable_secret'
sysctl: permission denied on key 'net.ipv6.conf.lo.stable_secret'
sysctl: permission denied on key 'vm.mmap_rnd_bits'
sysctl: permission denied on key 'vm.mmap_rnd_compat_bits'
sysctl: permission denied on key 'vm.stat_refresh'
```
You can use the following command In order to avoid these errors:
`sysctl kernel.shm_rmid_forced`
output 
`kernel.shm_rmid_forced = 1`

    command: ["bin/bash", "-c", "sleep INF"]
    securityContext:
      runAsUser: 2000 <1>
      runAsGroup: 3000 <2>
      allowPrivilegeEscalation: false <3>
      capabilities: <4>
        drop: ["ALL"]
  securityContext:
    runAsNonRoot: true <5> 
    seccompProfile: <6>
      type: RuntimeDefault
    sysctls:
    - name: kernel.shm_rmid_forced
      value: "1"
    - name: net.ipv4.ip_local_port_range
      value: "32770       60666"
    - name: net.ipv4.tcp_syncookies
      value: "0"
    - name: net.ipv4.ping_group_range
      value: "0           200000000"
----
<1> `runAsUser` controls which user ID the container is run with.
<2> `runAsGroup` controls which primary group ID the containers is run with.
<3> `allowPrivilegeEscalation` determines if a pod can request to allow privilege escalation. If unspecified, it defaults to true. This boolean directly controls whether the `no_new_privs` flag gets set on the container process.
<4> `capabilities` permit privileged actions without giving full root access. This policy ensures all capabilities are dropped from the pod. 
<5> `runAsNonRoot: true` requires that the container will run with a user with any UID other than 0. 

> <6> `RuntimeDefault` enables the default seccomp profile for a pod or container workload.

Reviewer 2: These are not related to the sysctls.
Do we want this described here?

Author: Peer review asked me to add callouts to be honest I am not sure the best approach. @kononovn stated there are changes in security in 4.11 but I could not find anything currently in docs or for upcoming 4.11 docs that cover these changes. Let me recycle to docs team. Is there nay specific details on these changes I think they change how pods run as privileged

Author: Maybe this is as you say too much information as unrelated to the feature in question. Was looking around there for a suitable cross ref I could put into the docs.
